src/gtest/test_proofs.cpp

```diff
@@ -0,0 +1,456 @@
+        curve_Fq("1234912378405347958234756902345768290345762348957605678245967234857634857676")
+    );
+
+    ASSERT_TRUE(x == x2.sqrt());
+    ASSERT_TRUE(nx == -x2.sqrt());
+    ASSERT_TRUE(x*x == x2);
+    ASSERT_TRUE(nx*nx == x2);
+}
+
+TEST(proofs, size_is_expected)
+{
+    ZCProof p;
+    CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
+    ss << p;
+
+    ASSERT_EQ(ss.size(), 296);
+}
+
+TEST(proofs, fq_serializes_properly)
+{
+    for (size_t i = 0; i < 1000; i++) {
+        curve_Fq e = curve_Fq::random_element();
+
+        Fq e2(e);
+
+        CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
+        ss << e2;
+
+        Fq e3;
+        ss >> e3;
+
+        curve_Fq e4 = e3.to_libsnark_fq<curve_Fq>();
+
+        ASSERT_TRUE(e == e4);
+    }
+}
+
+curve_Fq deserialize_from_test_vector(std::vector<unsigned char> tv)
+{
+    assert(tv.size() == 32);
+
+    uint256 expected_uint(tv);
+
+    CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
+    ss << expected_uint;
+
+    Fq e;
+    ss >> e;
+
+    curve_Fq e2 = e.to_libsnark_fq<curve_Fq>();
+
+    return e2;
+}
+
+TEST(proofs, fq_endianness)
+{
+    curve_Fq e = deserialize_from_test_vector({
+        0x30, 0x64, 0x4e, 0x72, 0xe1, 0x31, 0xa0, 0x29,
+        0xb8, 0x50, 0x45, 0xb6, 0x81, 0x81, 0x58, 0x5d,
+        0x97, 0x81, 0x6a, 0x91, 0x68, 0x71, 0xca, 0x8d,
+        0x3c, 0x20, 0x8c, 0x16, 0xd8, 0x7c, 0xfd, 0x46,
+    });
+
+    ASSERT_TRUE(e == curve_Fq("21888242871839275222246405745257275088696311157297823662689037894645226208582"));
+    ASSERT_TRUE(e != curve_Fq("21888242871839275222246405745257275088696311157297823662689037894645226208581"));
+
+    curve_Fq e2 = deserialize_from_test_vector({
+        0x30, 0x64, 0x4e, 0x72, 0xe1, 0x31, 0xa0, 0x29,
+        0xb7, 0x50, 0x45, 0xb6, 0x81, 0x81, 0x58, 0x5d,
+        0x97, 0x81, 0x6a, 0x91, 0x68, 0x71, 0xca, 0x8d,
+        0x3c, 0x20, 0x8c, 0x16, 0xd8, 0x7c, 0xfd, 0x46,
+    });
+
+    ASSERT_TRUE(e2 == curve_Fq("21888242871839275222221885816603420866962577604863418715751138068690288573766"));
+}
+
+TEST(proofs, fq_invalid)
+{
+    ASSERT_THROW(deserialize_from_test_vector({
+        0x30, 0x64, 0x4e, 0x72, 0xe1, 0x31, 0xa0, 0x29,
+        0xb8, 0x50, 0x45, 0xb6, 0x81, 0x81, 0x58, 0x5d,
+        0x97, 0x81, 0x6a, 0x91, 0x68, 0x71, 0xca, 0x8d,
+        0x3c, 0x20, 0x8c, 0x16, 0xd8, 0x7c, 0xfd, 0x47,
+    }), std::logic_error);
+
+    ASSERT_THROW(deserialize_from_test_vector({
+        0x30, 0x64, 0x4e, 0x72, 0xe1, 0x31, 0xa0, 0x29,
+        0xb8, 0x50, 0x45, 0xb6, 0x81, 0x81, 0x58, 0x5d,
+        0x97, 0x81, 0x6a, 0x91, 0x68, 0x71, 0xca, 0x8d,
+        0x3c, 0x20, 0x8c, 0x16, 0xd8, 0x7c, 0xfd, 0x48,
+    }), std::logic_error);
+
+    ASSERT_THROW(deserialize_from_test_vector({
+        0xff, 0x64, 0x4e, 0x72, 0xe1, 0x31, 0xa0, 0x29,
+        0xb8, 0x50, 0x45, 0xb6, 0x81, 0x81, 0x58, 0x5d,
+        0x97, 0x81, 0x6a, 0x91, 0x68, 0x71, 0xca, 0x8d,
+        0x3c, 0x20, 0x8c, 0x16, 0xd8, 0x7c, 0xfd, 0x46,
+    }), std::logic_error);
+}
```

Comment:
We don't have a class for Fq2, the only time we serialize/deserialize an Fq2 is with

```
        READWRITE(x0);
        READWRITE(x1);
```